api/apps/api/test/access-control/project-acl-get.e2e-spec.ts

```diff
@@ -0,0 +1,132 @@
+    projectId,
+  );
+  fixtures.ThenNoContentIsReturned(response);
+  response = await fixtures.WhenAddingANewOwnerToTheProjectAsOwner(projectId);
+  fixtures.ThenNoContentIsReturned(response);
+
+  response = await fixtures.WhenGettingProjectUsersAsOwner(projectId);
+  fixtures.ThenAllUsersinProjectAfterEveryTypeOfUserHasBeenAddedAreReturned(
+    response,
+  );
+});
+
+test(`add every type of user to a project as a project contributor`, async () => {
+  const projectId = await fixtures.GivenProjectWasCreated();
+  await fixtures.WhenAddingANewContributorToTheProjectAsOwner(projectId);
+
+  let response = await fixtures.WhenAddingANewViewerToTheProjectAsContributor(
+    projectId,
+  );
+  fixtures.ThenForbiddenIsReturned(response);
+  response = await fixtures.WhenAddingANewContributorToTheProjectAsContributor(
+    projectId,
+  );
+  fixtures.ThenForbiddenIsReturned(response);
+  response = await fixtures.WhenAddingANewOwnerToTheProjectAsContributor(
+    projectId,
+  );
+  fixtures.ThenForbiddenIsReturned(response);
+});
+
+test(`add every type of user to a project as a project viewer`, async () => {
+  const projectId = await fixtures.GivenProjectWasCreated();
+  await fixtures.WhenAddingANewContributorToTheProjectAsOwner(projectId);
+
+  let response = await fixtures.WhenAddingANewViewerToTheProjectAsViewer(
+    projectId,
+  );
+  fixtures.ThenForbiddenIsReturned(response);
+  response = await fixtures.WhenAddingANewContributorToTheProjectAsViewer(
+    projectId,
+  );
+  fixtures.ThenForbiddenIsReturned(response);
+  response = await fixtures.WhenAddingANewOwnerToTheProjectAsViewer(projectId);
+  fixtures.ThenForbiddenIsReturned(response);
+});
+
+test(`delete every type of user from the project as owner`, async () => {
+  const projectId = await fixtures.GivenProjectWasCreated();
+  await fixtures.WhenAddingANewViewerToTheProjectAsOwner(projectId);
+  await fixtures.WhenAddingANewContributorToTheProjectAsOwner(projectId);
+  await fixtures.WhenAddingANewOwnerToTheProjectAsOwner(projectId);
+  let revokeResponse = await fixtures.WhenRevokingAccessToViewerFromProjectAsOwner(
+    projectId,
+  );
+  fixtures.ThenNoContentIsReturned(revokeResponse);
+  revokeResponse = await fixtures.WhenRevokingAccessToContributorFromProjectAsOwner(
+    projectId,
+  );
+  fixtures.ThenNoContentIsReturned(revokeResponse);
+  revokeResponse = await fixtures.WhenRevokingAccessToOwnerFromProjectAsOwner(
+    projectId,
+  );
+  fixtures.ThenNoContentIsReturned(revokeResponse);
+  const response = await fixtures.WhenGettingProjectUsersAsOwner(projectId);
+  fixtures.ThenSingleOwnerUserInProjectIsReturned(response);
+});
+
+test(`delete last owner as the only owner`, async () => {
+  const projectId = await fixtures.GivenProjectWasCreated();
+  const revokeResponse = await fixtures.WhenRevokingAccessToLastOwnerFromProjectAsOwner(
+    projectId,
+  );
+  fixtures.ThenForbiddenIsReturned(revokeResponse);
```

Comment:
[minor] Is it forbidden operation due to ACL? (would consider BadRequest)

Reply:
I was wondering the same thing and I almost used BadRequest instead of Forbidden, but when I looked up the definition of 403 i found this (emphasis mine):

The request contained valid data and was understood by the server, but the server is refusing action. This may be due to the user not having the necessary permissions for a resource or needing an account of some sort, or **attempting a prohibited action** (e.g. creating a duplicate record where only one is allowed). This code is also typically used if the request provided authentication by answering the WWW-Authenticate header field challenge, but the server did not accept that authentication. The request should not be repeated.

As it was trying to do a prohibited action (removing the last owner from the project) I thought 403 was enough.